nrf52-hal-common/src/clocks.rs

```diff
@@ -1,61 +1,105 @@
 use crate::target::CLOCK;
-use crate::time::{Hertz, U32Ext};
+
+pub use crate::target::clock::lfclksrc::SRCW as LfOscSource;
 
 pub trait ClocksExt {
-    fn constrain(self) -> ClocksCfg;
+    fn constrain(self) -> Clocks<Internal, Internal, LfOscStopped>;
 }
 
-pub struct ClocksCfg {
-    pub hfclk: HFCLK,
-    pub lfclk: LFCLK,
-}
+pub const HFCLK_FREQ: u32 = 64_000_000;
+pub const LFCLK_FREQ: u32 =     32_768;
 
-pub struct HFCLK {
-    _0: ()
+pub struct Clocks<H, L, LSTAT> {
+    hfclk: H,
+    lfclk: L,
+    lfstat: LSTAT,
+    periph: CLOCK,
 }
 
-pub struct LFCLK {
-    _0: ()
-}
+pub struct Internal;
+pub struct ExternalOscillator;
+pub struct LfOscSynthesized;
 
-pub struct Clocks {
-    hfclk: Hertz,
-    lfclk: Hertz,
-}
+pub struct LfOscStarted;
+pub struct LfOscStopped;
 
-impl HFCLK {
-    // TODO: allow external clock selection?
-}
+impl<H, L, LSTAT> Clocks<H, L, LSTAT> {
+    pub fn enable_ext_hfosc(self) -> Clocks<ExternalOscillator, L, LSTAT> {
+        self.periph.tasks_hfclkstart.write(|w| unsafe { w.bits(1) });
+        Clocks {
+            hfclk: ExternalOscillator,
+            lfclk: self.lfclk,
+            lfstat: self.lfstat,
+            periph: self.periph,
+        }
+    }
 
-impl LFCLK {
-    // TODO: allow external clock selection? Calibration?
-}
+    pub fn disable_ext_hfosc(self) -> Clocks<Internal, L, LSTAT> {
+        self.periph.tasks_hfclkstop.write(|w| unsafe { w.bits(1) });
+        Clocks {
+            hfclk: Internal,
+            lfclk: self.lfclk,
+            lfstat: self.lfstat,
+            periph: self.periph,
+        }
+    }
 
-impl ClocksExt for CLOCK {
-    fn constrain(self) -> ClocksCfg {
-        ClocksCfg {
-            hfclk: HFCLK { _0: () },
-            lfclk: LFCLK { _0: () },
+    pub fn stop_lfclk(self) -> Clocks<H, L, LfOscStopped> {
+        self.periph.tasks_lfclkstop.write(|w| unsafe { w.bits(1) });
+        Clocks {
+            hfclk: self.hfclk,
+            lfclk: self.lfclk,
+            lfstat: LfOscStopped,
+            periph: self.periph,
         }
     }
-}
 
-impl ClocksCfg {
-    pub fn freeze(self) -> Clocks {
-        // TODO - this isn't very useful, can you actually change internal clock speeds?
+    pub fn start_lfclk(self) -> Clocks<H, L, LfOscStarted> {
+        self.periph.tasks_lfclkstart.write(|w| unsafe { w.bits(1) });
         Clocks {
-            hfclk: 64_000_000.hz(),
-            lfclk: 32_768.hz(),
+            hfclk: self.hfclk,
+            lfclk: self.lfclk,
+            lfstat: LfOscStarted,
+            periph: self.periph,
         }
     }
 }
 
-impl Clocks {
-    pub fn hfclk(&self) -> Hertz {
-        self.hfclk
+impl<H, L> Clocks<H, L, LfOscStopped> {
+    pub fn set_lfclk_src(self, src: LfOscSource, external: bool, bypass: bool) -> Clocks<H, Internal, LfOscStopped> {
```

Comment:
This signature is wrong. It always returns LFCLK: Internal, which is not true depending on `src`, I'll likely need to turn this into three methods to get the type signatures correct.